[JSC][GLib] Make types final in 2022 API

aperezdc opened the pull request:
#### a6d516900ca2f08da7ebcc3aa888b603d9681b93
<pre>
[JSC][GLib] Make types final in 2022 API
<a href="https://bugs.webkit.org/show_bug.cgi?id=252162">https://bugs.webkit.org/show_bug.cgi?id=252162</a>

Reviewed by Michael Catanzaro.

Switch the GLib JSC API to use WEBKIT_DEFINE_FINAL_TYPE in combination
with a new JSC_DECLARE_FINAL_TYPE macro for the public headers, modeled
after the WEBKIT_DECLARE_FINAL_TYPE macro from 260190@main. Applying the
same treatment to the JSC GLib types makes the 2022 API more coherent
and further reduces the amount of boilerplate code written manually, by
offloading the grunt work to the preprocessor macros.

The existing GENERATE_API_HEADERS macro get reused (and slightly
generalized) by moving it into a shared GLibMacros.cmake module, and
pressed into service to turn the JSC GLib API headers into templates.
All the autocleanups for JSC API types get ultimately defined by the
G_DECLARE_FINAL_TYPE macro, therefore the JSCAutocleanups.h header
installation get completely skipped with the new 2022 API.

* Source/JavaScriptCore/API/glib/JSCClass.cpp:
* Source/JavaScriptCore/API/glib/JSCClass.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCClass.h.
* Source/JavaScriptCore/API/glib/JSCContext.cpp:
* Source/JavaScriptCore/API/glib/JSCContext.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCContext.h.
* Source/JavaScriptCore/API/glib/JSCDefines.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCDefines.h.
* Source/JavaScriptCore/API/glib/JSCException.cpp:
* Source/JavaScriptCore/API/glib/JSCException.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCException.h.
* Source/JavaScriptCore/API/glib/JSCValue.cpp:
* Source/JavaScriptCore/API/glib/JSCValue.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCValue.h.
* Source/JavaScriptCore/API/glib/JSCVirtualMachine.cpp:
* Source/JavaScriptCore/API/glib/JSCVirtualMachine.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCVirtualMachine.h.
* Source/JavaScriptCore/API/glib/JSCWeakValue.cpp:
* Source/JavaScriptCore/API/glib/JSCWeakValue.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCWeakValue.h.
* Source/JavaScriptCore/API/glib/jsc.h.in: Renamed from Source/JavaScriptCore/API/glib/jsc.h.
* Source/JavaScriptCore/GLib.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/cmake/GLibMacros.cmake: Renamed from Source/WebKit/GLib.cmake.
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/260220@main">https://commits.webkit.org/260220@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf80a460cf16713bfef60b1e033a4b7960979763

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107577 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16633 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116732 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111471 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18058 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7947 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99743 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113335 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96823 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/41292 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/95544 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28449 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/96900 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9629 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/96312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/7641 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10295 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/96312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/15786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49383 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/105172 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7067 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11857 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/JavaScriptCore/API/glib/JSCClass.cpp:
```diff
@@ -65,17 +65,11 @@ typedef struct _JSCClassPrivate {
-    JSCClassPrivate* priv;
-};
-
 struct _JSCClassClass {
     GObjectClass parent_class;
 };
```
Why is this still here?

aperezdc replied:
Good catch, this should be inside a `#if !ENABLE(2022_GLIB_API)` guard. I'll do that before landing.

aperezdc committed:
[JSC][GLib] Make types final in 2022 API
https://bugs.webkit.org/show_bug.cgi?id=252162

Reviewed by Michael Catanzaro.

Switch the GLib JSC API to use WEBKIT_DEFINE_FINAL_TYPE in combination
with a new JSC_DECLARE_FINAL_TYPE macro for the public headers, modeled
after the WEBKIT_DECLARE_FINAL_TYPE macro from 260190@main. Applying the
same treatment to the JSC GLib types makes the 2022 API more coherent
and further reduces the amount of boilerplate code written manually, by
offloading the grunt work to the preprocessor macros.

The existing GENERATE_API_HEADERS macro get reused (and slightly
generalized) by moving it into a shared GLibMacros.cmake module, and
pressed into service to turn the JSC GLib API headers into templates.
All the autocleanups for JSC API types get ultimately defined by the
G_DECLARE_FINAL_TYPE macro, therefore the JSCAutocleanups.h header
installation get completely skipped with the new 2022 API.

* Source/JavaScriptCore/API/glib/JSCClass.cpp:
* Source/JavaScriptCore/API/glib/JSCClass.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCClass.h.
* Source/JavaScriptCore/API/glib/JSCContext.cpp:
* Source/JavaScriptCore/API/glib/JSCContext.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCContext.h.
* Source/JavaScriptCore/API/glib/JSCDefines.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCDefines.h.
* Source/JavaScriptCore/API/glib/JSCException.cpp:
* Source/JavaScriptCore/API/glib/JSCException.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCException.h.
* Source/JavaScriptCore/API/glib/JSCValue.cpp:
* Source/JavaScriptCore/API/glib/JSCValue.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCValue.h.
* Source/JavaScriptCore/API/glib/JSCVirtualMachine.cpp:
* Source/JavaScriptCore/API/glib/JSCVirtualMachine.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCVirtualMachine.h.
* Source/JavaScriptCore/API/glib/JSCWeakValue.cpp:
* Source/JavaScriptCore/API/glib/JSCWeakValue.h.in: Renamed from Source/JavaScriptCore/API/glib/JSCWeakValue.h.
* Source/JavaScriptCore/API/glib/jsc.h.in: Renamed from Source/JavaScriptCore/API/glib/jsc.h.
* Source/JavaScriptCore/GLib.cmake:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/cmake/GLibMacros.cmake: Renamed from Source/WebKit/GLib.cmake.
* Tools/TestWebKitAPI/PlatformGTK.cmake:
* Tools/TestWebKitAPI/PlatformWPE.cmake:

Canonical link: https://commits.webkit.org/260220@main
